### PR TITLE
Reduce OrderBy overheads for default comparer

### PR DESCRIPTION
**Pull request by stephentoub**
If OrderBy is used with a value type key and and a default or no comparer is specified, we can use `Comparer<TKey>.Default.Compare` to avoid the interface dispatch.  This optimization can be taken further, but for now it handles a common set of cases. (We can subsequently look into a variety of improvements, such as a cheaper way to implement a stable sort, doing the sort in-place in the buffered output when conditions are ammenable, extending these checks for using the default comparer into chained OrderBy/ThenBy cases, etc.)

```C#
[Params(10, 100, 1_000, 1_000_000)]
public int Length { get; set; }

private int[] _values;

[GlobalSetup]
public void Setup() => _values = Enumerable.Range(0, Length).Reverse().ToArray();

[Benchmark]
public int[] OrderByToArray() => _values.OrderBy(i => i).ToArray();
```

|         Method |         Toolchain |  Length |             Mean | Ratio |
|--------------- |------------------ |-------- |-----------------:|------:|
| OrderByToArray | \main\corerun.exe |      10 |         418.8 ns |  1.00 |
| OrderByToArray |   \pr\corerun.exe |      10 |         314.6 ns |  0.75 |
|                |                   |         |                  |       |
| OrderByToArray | \main\corerun.exe |     100 |       4,291.3 ns |  1.00 |
| OrderByToArray |   \pr\corerun.exe |     100 |       2,680.5 ns |  0.63 |
|                |                   |         |                  |       |
| OrderByToArray | \main\corerun.exe |    1000 |      71,427.6 ns |  1.00 |
| OrderByToArray |   \pr\corerun.exe |    1000 |      40,393.7 ns |  0.57 |
|                |                   |         |                  |       |
| OrderByToArray | \main\corerun.exe | 1000000 | 180,572,845.2 ns |  1.00 |
| OrderByToArray |   \pr\corerun.exe | 1000000 |  96,260,728.2 ns |  0.53 |

### CODE REVIEW

File: src/libraries/System.Linq/src/System/Linq/OrderedEnumerable.cs [hunk 1]
```diff
@@ -359,10 +359,43 @@ internal override int CompareAnyKeys(int index1, int index2)
...
+                c == 0 ? index1 - index2 : // ensure stability of sort
+                c;
+        }
+
         private int CompareKeys(int index1, int index2) => index1 == index2 ? 0 : CompareAnyKeys(index1, index2);
 
         protected override void QuickSort(int[] keys, int lo, int hi) =>
-            new Span<int>(keys, lo, hi - lo + 1).Sort(CompareAnyKeys);
+            new Span<int>(keys, lo, hi - lo + 1).Sort(
+                !typeof(TKey).IsValueType || _next is not null || _comparer != Comparer<TKey>.Default ? new Comparison<int>(CompareAnyKeys) :
```
Comment (eiriktsarpalis) [inline src/libraries/System.Linq/src/System/Linq/OrderedEnumerable.cs:396]: Nit: I suspect lifting the comparison resolution expression to a variable binding might improve legibility.

File: src/libraries/System.Linq/src/System/Linq/OrderedEnumerable.cs [hunk 2]
```diff
@@ -359,10 +359,43 @@ internal override int CompareAnyKeys(int index1, int index2)
...
+                _descending ? new Comparison<int>(CompareAnyKeys_DefaultComparer_NoNext_Descending) :
+                new Comparison<int>(CompareAnyKeys_DefaultComparer_NoNext_Ascending));
```
Comment (eiriktsarpalis) [inline src/libraries/System.Linq/src/System/Linq/OrderedEnumerable.cs:398]: I take it that the QuickSort method is only called once in the lifetime of this object and we're fine with allocating a new delegate every time?

File: src/libraries/System.Linq/src/System/Linq/OrderedEnumerable.cs [hunk 3]
```diff
@@ -359,10 +359,43 @@ internal override int CompareAnyKeys(int index1, int index2)
...
+
+            int c = Comparer<TKey>.Default.Compare(keys[index1], keys[index2]);
+            return
+                c == 0 ? index1 - index2 : // ensure stability of sort
+                c;
+        }
+
+        private int CompareAnyKeys_DefaultComparer_NoNext_Descending(int index1, int index2)
+        {
+            Debug.Assert(_comparer == Comparer<TKey>.Default);
```
Comment (eiriktsarpalis) [inline src/libraries/System.Linq/src/System/Linq/OrderedEnumerable.cs:379]:
```suggestion
            Debug.Assert(typeof(TKey).IsValueType);
            Debug.Assert(_comparer == Comparer<TKey>.Default);
```

File: src/libraries/System.Linq/src/System/Linq/OrderedEnumerable.cs [hunk 4]
```diff
@@ -359,10 +359,43 @@ internal override int CompareAnyKeys(int index1, int index2)
             return (_descending != (c > 0)) ? 1 : -1;
         }
 
+        private int CompareAnyKeys_DefaultComparer_NoNext_Ascending(int index1, int index2)
+        {
+            Debug.Assert(_comparer == Comparer<TKey>.Default);
```
Comment (eiriktsarpalis) [inline src/libraries/System.Linq/src/System/Linq/OrderedEnumerable.cs:364]:
```suggestion
            Debug.Assert(typeof(TKey).IsValueType);
            Debug.Assert(_comparer == Comparer<TKey>.Default);
```

**Review by eiriktsarpalis — APPROVED** [review]
Thanks

File: src/libraries/System.Linq/src/System/Linq/OrderedEnumerable.cs (hunk 2 above)
Comment (stephentoub) [reply src/libraries/System.Linq/src/System/Linq/OrderedEnumerable.cs:398]: It is. "fine" is a strong term :-) We'd prefer not to, but we don't have a better option right now. Note this was allocating a delegate already; I just changed the syntax to be explicit about it.

**Pull request merged**